src/index/IndexMetaData.h

```diff
@@ -18,175 +25,49 @@ using std::vector;
-  // If the lhs does not exists at all, this will only be clear after reading
-  // said block.
-  pair<off_t, size_t> getBlockStartAndNofBytesForLhs(Id lhs) const;
-
-  // Gets the block after the one returned by getBlockStartAndNofBytesForLhs.
-  // This is necessary for finding rhs upper bounds for the last item in a
-  // block.
-  // If this is equal to the block returned by getBlockStartAndNofBytesForLhs,
-  // it means it is the last block and the offsetAfter can be used.
-  pair<off_t, size_t> getFollowBlockForLhs(Id lhs) const;
-
-  off_t _startRhs;
-  off_t _offsetAfter;
-  vector<BlockMetaData> _blocks;
-};
-
-inline ad_utility::File& operator<<(ad_utility::File& f,
-                                    const BlockBasedRelationMetaData& rmd) {
-  f.write(&rmd._startRhs, sizeof(rmd._startRhs));
-  f.write(&rmd._offsetAfter, sizeof(rmd._offsetAfter));
-  auto nofBlocks = rmd._blocks.size();
-  f.write(&nofBlocks, sizeof(nofBlocks));
-  f.write(rmd._blocks.data(), nofBlocks * sizeof(BlockMetaData));
-  return f;
-}
-
-class RelationMetaData {
- public:
-  explicit RelationMetaData(const FullRelationMetaData& rmdPairs)
-      : _rmdPairs(rmdPairs), _rmdBlocks(nullptr) {}
-
-  off_t getStartOfLhs() const { return _rmdPairs.getStartOfLhs(); }
-
-  size_t getNofBytesForFulltextIndex() const {
-    return _rmdPairs.getNofBytesForFulltextIndex();
-  }
-
-  // Returns true if there is exactly one RHS for each LHS in the relation
-  bool isFunctional() const { return _rmdPairs.isFunctional(); }
-
-  bool hasBlocks() const { return _rmdPairs.hasBlocks(); }
-
-  size_t getNofElements() const { return _rmdPairs.getNofElements(); }
-
-  uint8_t getCol1LogMultiplicity() const {
-    return _rmdPairs.getCol1LogMultiplicity();
-  }
-
-  uint8_t getCol2LogMultiplicity() const {
-    return _rmdPairs.getCol2LogMultiplicity();
-  }
-
-  const FullRelationMetaData& _rmdPairs;
-  const BlockBasedRelationMetaData* _rmdBlocks;
-};
-
-inline ad_utility::File& operator<<(ad_utility::File& f,
-                                    const RelationMetaData& rmd) {
-  f << rmd._rmdPairs << *rmd._rmdBlocks;
-  return f;
-}
-
-class IndexMetaData {
- public:
-  IndexMetaData();
-
+  // some MapTypes (the dense ones using stxxl or mmap) require additional calls
+  // to setup() before being fully initialized
+  IndexMetaDataTemplated() = default;
+
+  // pass all arguments that are needed for initialization to the underlying
+  // implementation of MapTyep _data
+  template<typename... dataArgs>
+  void setup(dataArgs... args) {_data.setup(args...);}
+
+  // already Existing means we do not need to at rmd to _data
+  // if bool alreadyExisting is set, we will not add argument rmd 
+  // but assume that is already is contained in _data (for persistent metaData
+  // implementations
   void add(const FullRelationMetaData& rmd,
-           const BlockBasedRelationMetaData& bRmd);
+           const BlockBasedRelationMetaData& bRmd, bool alreadyExisting = false);
```

Comment:
This formatting looks wrong, have you set up `clang-format` it should automatically pick up the `.clang-format` from the project root